Support geo-replication functionality

xaniasd opened the pull request:
With these changes, heketi can support geo-replication features of gluster in their basic form. The intention was to stay close to the glusterfs syntax as described [here](http://gluster.readthedocs.io/en/latest/Administrator%20Guide/Geo%20Replication/). At this point, nothing is stored in the database.

Relevant changes:

* executors extended to support the various geo-replication commands
* endpoints defined on the server-side
* heketi-cli commands added

examples of heketi-cli commands:

```
$ heketi-cli georep status #returns the status information from the first node of the first cluster
$ heketi-cli volume georep --slave-host=blah --slave-volume=23423423 create 886a86a868711bef83001
$ heketi-cli volume georep --slave-host=blah --slave-volume=23423423 config 886a86a868711bef83001 --sync-jobs 2 --ssh-port 2222
$ heketi-cli volume georep --slave-host=blah --slave-volume=23423423 start 886a86a868711bef83001
$ heketi-cli volume georep status 886a86a868711bef83001
$ heketi-cli volume georep --slave-host=blah --slave-volume=23423423 stop 886a86a868711bef83001
$ heketi-cli volume georep --slave-host=blah --slave-volume=23423423 delete 886a86a868711bef83001
```

obnoxxx commented on client/cli/go/cmds/volume.go:
```diff
@@ -263,7 +264,7 @@ var volumeExpandCommand = &cobra.Command{
 	Example: `  * Add 10GB to a volume
     $ heketi-cli volume expand --volume=60d46d518074b13a04ce1022c8c7193c --expand-size=10
-`,
+	`,
```
spacing change that does not belong to this commit

obnoxxx commented on client/cli/go/cmds/volume.go:
```diff
@@ -112,28 +113,28 @@ var volumeCreateCommand = &cobra.Command{
+	--disperse-data=8 --redundancy=3
+	
+	* Create a 100GB distributed volume which supports performance related volume options.
+	$ heketi-cli volume create --size=100 --durability=none --gluster-volume-options="performance.rda-cache-limit 10MB","performance.nl-cache-positive-entry no"
```
This whole block seems to be a reformatting of white spaces.
This is not related to this commit and should hence not be part of it.

obnoxxx commented on executors/executor.go:
```diff
@@ -9,7 +9,9 @@
-import "encoding/xml"
+import (
+	"encoding/xml"
+)
```
unrelated reformatting. either put it in a separate commit if there is a good reason, or otherwise skip it, but don't sneak it in.

obnoxxx commented:
Hi @xaniasd, thanks a lot for your contribution!

I have looked at the patches a little bit, and they seem reather clean (even though the second patch "implement geo-rep related functionality" should probably be merged with the first one "Working create command", since it re-does several things that the first one did differently.

But I have to say that I currently fail to fully get the point of the change: If this does not currently store anything in the DB, then it's merely a frontend to gluster cli, and after the setup of a geo-rep, heketi will not have any knowledge of its existence any more. So what's the extra value this adds? Is it consistency of interface?

And one concrete question. In the description of the PR you give the example
```
heketi-cli volume georep --slave-host=blah --slave-volume=23423423 create 886a86a868711bef83001
```
What is `886a86a868711bef83001` here? It is the (master) volume id? Or should the user
invent an id for the replication?

xaniasd commented:
Hi @obnoxxx 

in my use case we use heketi to remotely manage glusterfs volumes. I would like to use the same channel to manage geo-replication as well, instead of opening up a new one to access the gluster cli. This serves our purpose fine as a gluster cli frontend. I considered storing geo-replication related state in the database, but had no concrete usage of that and got quite some mileage without it. 

This is furthermore a first attempt to managing geo-replication through Heketi and I'm sure we can improve on it. We could store for instance an extra flag in Heketi (if geo-replication is configured or not) to save calls to gluster, or other improvements, I'm open to suggestions :)

About the second question: I tried to keep it in line with the other volume sub-commands, so the argument is the id of the master volume.

And about your very first suggestion, should I make a new PR with the commits squashed?

obnoxxx commented:
@xaniasd thanks for the eplanations.
I'll need a little more time to think about it.
@raghavendra-talur , @lpabon , @humblec , please share any thoughts about adding this API for geo-rep!

And for the cosmetic: 
Yes, I meant squashing thos patches that belong together.
And I'm perfectly fine with you force-pushing over this PR.
Please don't merge other stuff into your branch and push. This is confusing to me.
But you can rebase on top of an updated origin/master, and also use `git rebase --interactive origin/master` in combination with `git push --force ...` to update this PR in place. It will/might lose my old comments, but I guess this is ok.
If you have any more questions, please ask.

humblec commented:
@xaniasd First of all thanks for this patch! One quick question from my side, have you got a chance to check the behaviour if there is a `restart` of heketi service in between? I am asking this explicitly because of lack of state store in heketi db.

xaniasd commented:
@obnoxxx ok, I'll do that. On a sidenote, I have fixed some minor bugs after more thorough tests, and added some unit tests. I'll submit those patches as a separate PR then.

@humblec The lack of state should make it more robust. Besides, all geo-rep actions are largely "atomic" as opposed to e.g. creating a volume that comprise multiple steps. So I suppose a restart wouldn't hurt much more than a failed HTTP call, right? But just to be on the same page, do you mean issuing a geo-rep command (e.g. from the heketi client) and then restarting the service, or something else? If you have any test cases in mind that could lead to trouble, I'd be happy to test them out for you.

xaniasd commented:
ping @obnoxxx, @humblec is there something else missing from the PR?

obnoxxx commented on apps/glusterfs/app_georep.go:
```diff
@@ -1,3 +1,11 @@
+//
+// Copyright (c) 2015 The heketi Authors
```
2017

obnoxxx commented on apps/glusterfs/georep.go:
```diff
@@ -49,6 +49,15 @@ func (v *VolumeEntry) GeoReplicationAction(db *bolt.DB,
 }
 
+//
+// Copyright (c) 2015 The heketi Authors
```
2017

obnoxxx commented on client/cli/go/cmds/georep.go:
```diff
@@ -1,3 +1,12 @@
+//
+// Copyright (c) 2015 The heketi Authors
```
2017

obnoxxx commented on executors/sshexec/georep.go:
```diff
@@ -1,3 +1,12 @@
+//
+// Copyright (c) 2015 The heketi Authors
```
2017

obnoxxx commented:
@xaniasd thanks a lot for the updates to the PR!
This is highly appreciated!

I have added a few more review comments.

For the question what is needed: I need the time to sit down and really think about this new API and test the code. We should discuss if this is exactly the API extension we need / want.

Ideally we might want a design document description in doc/design for this new feature.
Hence putting up the "do not merge" label just for now: it's a sign for me that I still need some more discussion around this, even if the patch itself starts looking really good.

Keep it up!

:-)

xaniasd commented:
@obnoxxx  I don't see the point of working on a good patch if it's not going to be merged eventually. So let's focus on the discussion part. What are your concrete reservations about the change? Is there some implementation detail that should be different or does it not agree with the general approach?

This is not a breaking change and I have taken pains to keep it in line with the rest of the API. I understand that you might not have a direct need of this feature, but I have clearly described my use case and I hope as a user of heketi to belong to the collective "we" as much as anyone else.

Once we're on the same page, I'll be happy to add documentation and tests.

kannanvr commented:
@xaniasd , When this feature will be available. I wanted this feature on heketi.

abessifi commented:
Is this still a WIP ?
Do geo-replication CNS-to-CNS is supported in Heketi ?

phlogistonjohn commented:
Closing PRs older than 2018. No expectation this will be continued. Feel free to reopen if things change.
